File: MessagesKit/FMResultSet+Utils.m
//
//  FMResultSet+Utils.m
//  MessagesKit
//
//  Created by Kevin Wooten on 4/20/16.
//  Copyright © 2016 reTXT Labs LLC. All rights reserved.
//

#import "FMResultSet+Utils.h"

#import "Messages+Exts.h"
#import "ExternalFileDataReference.h"
#import "DataReferences.h"



@interface DataReferenceInflater : NSObject <NSKeyedUnarchiverDelegate>

@property(nonatomic, retain) DBManager *dbManager;

-(instancetype) initWithDBManager:(DBManager *)dbManager;

@end


@interface DataReferenceDeflater : NSObject <NSKeyedArchiverDelegate>

@property(nonatomic, retain) DBManager *dbManager;

-(instancetype) initWithDBManager:(DBManager *)dbManager;

@end




@implementation FMResultSet (Model)

-(NSURL *) URLForColumn:(NSString *)columnName
{
  NSString *val = [self stringForColumn:columnName];
  return val ? [NSURL URLWithString:val] : nil;
}

-(NSURL *) URLForColumnIndex:(int)columnIdx
{
  NSString *val = [self stringForColumnIndex:columnIdx];
  return val ? [NSURL URLWithString:val] : nil;
}

-(Id *) idForColumn:(NSString *)columnName
{
  NSData *val = [self dataForColumn:columnName];
  return val ? [Id idWithData:val] : nil;
}

-(Id *) idForColumnIndex:(int)columnIdx
{
  NSData *val = [self dataForColumnIndex:columnIdx];
  return val ? [Id idWithData:val] : nil;
}

-(CGSize) sizeForColumn:(NSString *)columnName
{
  NSString *val = [self stringForColumn:columnName];
  return val ? CGSizeFromString(val) : CGSizeZero;
}

-(CGSize) sizeForColumnIndex:(int)columnIdx
{
  NSString *val = [self stringForColumnIndex:columnIdx];
  return val ? CGSizeFromString(val) : CGSizeZero;
}

-(id<DataReference>) dataReferenceForColumn:(NSString *)columnName usingDBManager:(DBManager *)dbManager
{
  NSData *data = [self dataForColumn:columnName];
  if (!data) {
    return nil;
  }
  
  id<NSKeyedUnarchiverDelegate> uaDelegate = [DataReferenceInflater.alloc initWithDBManager:dbManager];
  NSKeyedUnarchiver *ua = [NSKeyedUnarchiver.alloc initForReadingWithData:data];
  ua.delegate = uaDelegate;
  
  return [ua decodeObjectForKey:NSKeyedArchiveRootObjectKey];
}

-(id<DataReference>) dataReferenceForColumnIndex:(int)columnIndex usingDBManager:(DBManager *)dbManager
{
  NSData *data = [self dataForColumnIndex:columnIndex];
  if (!data) {
    return nil;
  }
  
  id<NSKeyedUnarchiverDelegate> uaDelegate = [DataReferenceInflater.alloc initWithDBManager:dbManager];
  NSKeyedUnarchiver *ua = [NSKeyedUnarchiver.alloc initForReadingWithData:data];
  ua.delegate = uaDelegate;
  
  return [ua decodeObjectForKey:NSKeyedArchiveRootObjectKey];
}

-(id) nillableObjectForColumnIndex:(int)columnIndex
{
  id val = [self objectForColumnIndex:columnIndex];
  if (val == [NSNull null]) {
    return nil;
  }
  return val;
}

@end




@implementation DataReferenceInflater

-(instancetype) initWithDBManager:(DBManager *)dbManager
{
  self = [self init];
  if (self) {
    self.dbManager = dbManager;
  }
  return self;
}

-(id)unarchiver:(NSKeyedUnarchiver *)unarchiver didDecodeObject:(id)object
{
  if ([object isKindOfClass:ExternalFileDataReference.class]) {
    ExternalFileDataReference *externalFileRef = (id)object;
    return [ExternalFileDataReference.alloc initWithDBManager:self.dbManager
                                                     fileName:externalFileRef.fileName];
  }
  return object;
}

@end


@implementation DataReferenceDeflater

-(instancetype) initWithDBManager:(DBManager *)dbManager
{
  self = [self init];
  if (self) {
    self.dbManager = dbManager;
  }
  return self;
}

@end

File: Messages/FMResultSet+Utils.swift
//
//  FMResultSet+Utils.swift
//  Messages
//
//  Created by Kevin Wooten on 4/24/16.
//  Copyright © 2016 reTXT Labs LLC. All rights reserved.
//

import Foundation


extension FMResultSet {

  @objc public func dataReferenceForColumn(columnName: String, forOwner owner: String, usingDB db: RTDBManager) -> DataReference? {
    
    guard let data = dataForColumn(columnName) else {
      return nil
    }
    
    return dataReferenceFromData(data, owner: owner, db: db)
  }
  
  @objc public func dataReferenceForColumnIndex(columnIndex: Int32, forOwner owner: String, usingDB db: RTDBManager) -> DataReference? {
    
    guard let data = dataForColumnIndex(columnIndex) else {
      return nil
    }
    
    return dataReferenceFromData(data, owner: owner, db: db)
  }
  
  private func dataReferenceFromData(data: NSData, owner: String, db: RTDBManager) -> DataReference? {
    
    let ua = NSKeyedUnarchiver(forReadingWithData: data)
    
    let uaDelegate = DataReferenceInflater(db: db, owner: owner)
    ua.delegate = uaDelegate
    
    return ua.decodeObjectForKey(NSKeyedArchiveRootObjectKey) as? DataReference
  }
  
}


class DataReferenceDeflater: NSObject, NSKeyedUnarchiverDelegate {
  
  private let db: RTDBManager
  private let owner: String
  
  init(db: RTDBManager, owner: String) {
    self.db = db
    self.owner = owner
  }
  
  
}

class DataReferenceInflater: NSObject, NSKeyedUnarchiverDelegate {
  
  private let db: RTDBManager
  private let owner: String
  
  init(db: RTDBManager, owner: String) {
    self.db = db
    self.owner = owner
  }
  
  func unarchiver(unarchiver: NSKeyedUnarchiver, didDecodeObject object: AnyObject?) -> AnyObject? {
    
    if let blobRef = object as? BlobDataReference {
      blobRef.db = db
    }
    
    return object
  }
  
}
